<a href="https://colab.research.google.com/github/kenltho2/Num-Meth23/blob/main/project7question2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [32]:
import math
import pandas as pd
import numpy as np
import matplotlib as plt
import scipy

**For this portion of the project we will be looking at Simpson's Rule. Simpson's Rule approximates definite integrals by interpolating a polynomial to fit the integrand based on given points and then calculates areas under sections of the integrand based on the interpolation. The formula for Simpson's Rule can be derived from the polynomial to give us an integral estimate of $
\begin{equation*}
\int_{a}^{b}f(x)\, dx \approx \frac{b-a}{6} \big[f(a) + 4f(\frac{a+b}{2}) + f(b) \big]
\end{equation*}
$.**

In [39]:
def simpson(a,b,n):
  h = (b-a)/(n-1)
  x = np.linspace(a,b,n)
  f = np.exp(x)
  est = ((h/3) * (f[0]+ 4*sum(f[i] for i in range(1,n-1,2)) + 2*sum(f[j] for j in range(0,n-2,2)) + f[n-1]))
  return est
estimate = simpson(1,3,15)
estimate

17.62617916620393

In [41]:
def F(x):
  return math.exp(x)
real = scipy.integrate.quad(F,1,3)
real[0]

17.367255094728623

In [47]:
def true_error(true,approx):
  return true - approx
true_error(real[0],estimate)

-0.25892407147530605

In [45]:
def rel_true_error(true, approx):
 E = ((true - approx) / true) * 100
 return E
rel_true_error(real[0],estimate)

-1.4908750407765685

**Based on the error calculations, Simpson's Rule slightly underestimates the function $e^x$ by about $-1.49\%$. I would use Simpson's Rule to calculate integrals of a function that undulates such as $sin(x)$ and $cos(x)$ because polynomial interpolation mimics functions that undulate more intuitively and not so much functions that strictly increase like $e^x$, however it still produces a fairly accurate estimate.**